### PR TITLE
Helpers for modifying request headers as middleware

### PR DESCRIPTION
**Pull request by bitemyapp**
Before submitting your PR, check that you've:

- [X] Bumped the version number

Yep, minor bump because it's a backwards compatible addition.

- [X] Documented new APIs with [Haddock markup](https://www.haskell.org/haddock/doc/html/index.html)
- [x] Added [`@since` declarations](http://haskell-haddock.readthedocs.io/en/latest/markup.html#since) to the Haddock

After submitting your PR:

- [x] Update the Changelog.md file with a link to your PR
- [ ] Check that CI passes (or if it fails, for reasons unrelated to your change, like CI timeouts)

### CODE REVIEW

File: wai/Network/Wai.hs [hunk 1]
```diff
@@ -303,6 +306,10 @@ defaultRequest = Request
 type Middleware = Application -> Application
 
 
+-- | apply a function that modifies a request as a 'Middleware'
+modifyRequest :: (Request -> Request) -> Middleware
+modifyRequest f app req respond = app (f req) $ respond
```
Comment (snoyberg) [inline wai/Network/Wai.hs]: You can eta reduce and remove the `respond`.

File: wai/Network/Wai.hs [hunk 2]
```diff
@@ -303,6 +306,10 @@ defaultRequest = Request
 type Middleware = Application -> Application
 
 
+-- | apply a function that modifies a request as a 'Middleware'
```
Comment (snoyberg) [inline wai/Network/Wai.hs]: Can you add a `@since` comment here?

File: wai/Network/Wai.hs [hunk 3]
```diff
@@ -344,3 +351,9 @@ lazyRequestBody req =
             else do
                 bss <- loop
                 return $ LI.Chunk bs bss
+
+-- | Apply the provided function to the request header list of the Request.
```
Comment (snoyberg) [inline wai/Network/Wai.hs]: Perhaps add a comment about this being intended to work with `modifyRequest`?

**Comment by Vlix** [comment]
Added to `wai-3.2.4`

**Pull request closed without merging**